src/protocol/payload_factory.rs

```diff
@@ -1,29 +1,80 @@
-use crate::protocol::codecs::payload::Payload;
-
-/// A factory for creating payloads.
-#[derive(Clone)]
-pub struct PayloadFactory {
-    payload: Payload,
-}
-
-impl PayloadFactory {
-    /// Create a new payload factory using specified payload as a template.
-    pub fn new(payload: Payload) -> Self {
-        Self { payload }
-    }
-
-    // TODO[asmie]: Add generate() method to start generating payloads (first message).
-    // TODO[asmie]: Re-think API of the PayloadFactory to handle all generic cases (like creating custom payloads with custom fields).
-
-    /// Create a new payload with the same type as the template. If there is a need to
-    /// change any payload fields it's done here.
-    pub fn generate_next(&mut self) -> Payload {
-        // For now, we're incrementing nonce in UniEnsBlockReq and not change other
-        // type of messages.
-        if let Payload::UniEnsBlockReq(message) = &mut self.payload {
-            message.nonce += 1;
-        }
-
-        self.payload.clone()
-    }
-}
+use crate::protocol::codecs::payload::Payload;
+
+/// A factory for creating payloads.
+#[derive(Clone)]
+pub struct PayloadFactory {
+    payload: Payload,
+    customize_payload: fn(&mut Payload) -> (),
+    pre_generated_cache: Vec<Payload>,
+}
+
+impl PayloadFactory {
+    /// Create a new payload factory using specified payload as a template and a function to customize
+    /// the payload. If no customization is needed, just pass None and use default customizer.
+    pub fn new(payload: Payload, customize_payload: Option<fn(&mut Payload) -> ()>) -> Self {
+        let default_customize_payload = |msg: &mut Payload| match msg {
+            Payload::UniEnsBlockReq(message) => {
+                message.nonce += 1;
+            }
+            Payload::UniCatchupReq(message) => {
+                message.nonce += 1;
+            }
+            _ => {}
+        };
+
+        Self {
+            payload,
+            customize_payload: customize_payload.unwrap_or(default_customize_payload),
+            pre_generated_cache: Vec::new(),
+        }
+    }
+
+    /// Create a new payload with the same type as the template. If there is a need to
+    /// change any payload fields customizer is run.
+    pub fn generate_next(&mut self) -> Payload {
+        (self.customize_payload)(&mut self.payload);
+        self.payload.clone()
+    }
+
+    /// Create a new payload with the same type as the template and store it in the cache.
+    pub fn pre_generate_payloads_cache(&mut self, count: usize) {
+        self.pre_generated_cache = self.generate_payloads(count);
+    }
+
+    /// Get vector of payloads reference from the cache.
+    pub fn get_pre_generated_payload_cache(&self) -> &Vec<Payload> {
```

Comment:
hmm, do we even need this function? `generate_next` could handle this internally:
- either generate a new payload
- or return a pre generated payload from the cache if caches is not empty (maybe crucial for performance tests)

doing it like that makes usage PayloadFactory API simpler:
user does basically:
```rust
PayloadFactory::new(...)  // call once
PayloadFactory::generate_next() // call whenever needed
```
or optionally (for performance tests e.g.):
```rust
PayloadFactory::new(...)  // call once
PayloadFactory::pre_generate_payloads_cache(big_n)  // call once
PayloadFactory::generate_next() // call whenever needed
```

Reply:
Ok, let's start without the cache API and see whether it will be needed.